### ISSUE TITLE
New Behaviour of add command

### ISSUE THREAD

**Issue by kevina** [ipfs/go-ipfs#3342]
Adding multiple files at the same time with `ipfs add` is basically broken.  #3258 and #3300 basically propose different solutions.

#3258 adds and pins each entry specified on the command line individually without creating a common mfs-root, but it does not work with directories.

#3300 makes add behave more like tar and also adds the directories as given on the command line and then adds everything under a single pin and mfs-root

**Comment by whyrusleeping** [comment]
Alright, so heres what i'm thinking.

```
ipfs add FILE

# should produce output of

added HASH FILE
# and HASH should be pinned recursively
```

```
ipfs add FOO/FILE

added HASH FOO/FILE
# and HASH should be pinned recursively
```

```
ipfs add foo/bar baz/qux

added <hash> foo/bar
added <hash> baz/qux
# and both should be pinned recursively (two pins)
```

```
ipfs add -w foo/bar baz/qux

added <hash> foo/bar
added <hash> baz/qux
added <root hash>
# and <root hash> should be pinned recursively (one pin)
```

We can accomplish this with a combination of the 'semi-full-root path' on #3300, and the IsRoot functionality i was abandoning. My current idea is to send the isRoot over, and have the adder keep a map of which nodes are 'roots'. Then on the outputDirs step, we can stop printing when we reach a node thats marked as a root.

**Comment by kevina** [comment]
So `-w` enables the tar like behavior, that sounds good to me?

**Comment by whyrusleeping** [comment]
@kevina Yeah. Sound good?

**Comment by kevina** [comment]
What about `-r`?

**Comment by whyrusleeping** [comment]
`-r` shouldnt change any of the above. The file examples could all be files or directories and the above behaviours should remain the same

**Comment by kevina** [comment]
Yeah that sounds good.  Do you think you can implement this quickly or do you want me to?

**Comment by whyrusleeping** [comment]
@kevina If you like, I can go ahead and do this. Depending on how distracted i get i might be able to have it done tomorrow.

If you want to take it I have more than enough other things to do too, so whatever works

**Comment by kevina** [comment]
I can take it no problem.  The only thing I could use help with is passing IsRoot() through the API.  If you had that working before pointing me to that code could save me some time.

**Comment by kevina** [comment]
@whyrusleeping, on second thought I am not 100% sure what you had in mind, if going to do what I think it will it will still involve creating a whole bunch of useless directories when adding files in my use case:  my scripts will add of unrelated files bunch of files based on the absolute path.  I would prefer a simpler code path that just adds files and doesn't try to add any sort of directories.  But it should work well enough and I can live with it.

It will likely be easier if you implement this and then I will see if it works for me.

**Comment by kevina** [comment]
I like the interface though, just not how I think you want it implemented.
